packages/go_router/README.md

```diff
@@ -119,6 +119,40 @@ GoRouter(
 );
 ```
 
+## Redirection
+
+You can use redirection to prevent user from visiting a specific page. In
+go_router, redirection can be asynchronous.
+
+```dart
+GoRouter(
+  ...
+  redirect: (context, state) async {
+    if (await LoginService.of(context).isLoggedIn) {
+      return state.location;
+    }
+    return '/login';
+  },
+);
+```
+If the code depends on [BuildContext](https://api.flutter.dev/flutter/widgets/BuildContext-class.html)
+through the [dependOnInheritedWidgetOfExactType](https://api.flutter.dev/flutter/widgets/BuildContext/dependOnInheritedWidgetOfExactType.html)
+(which is how `of` method is usually implemented), the redirect will be called every time the [InheritedWidget](https://api.flutter.dev/flutter/widgets/InheritedWidget-class.html)
+updated.
+
+### Top-level redirect
+
+The [GoRouter.redirect](https://pub.dev/documentation/go_router/latest/go_router/GoRouter-class.html)
+will be called for every navigation regardless of which GoRoute it matches. It
+always takes priority over route-level redirect.
+
+### Route-level redirect
+
+The [GoRoute.redirect](https://pub.dev/documentation/go_router/latest/go_router/GoRoute/redirect.html)
+will only be called if the route has matched the GoRoute. If there are multiple
+GoRoute matches, e.g. GoRoute with sub-routes. The parent route redirect takes
+priority over sub-routes' redirect.
```

Comment:
Should we reiterate that the top-level redirect will take precedence over the `GoRoute`'s redirect here? This would be repetitive, but, useful for anyone that skipped the top-level redirect section.